src/AddressRegistry.Api.Legacy/AddressMatch/Matching/MunicipalityMatcher.cs

```diff
@@ -6,112 +6,155 @@ namespace AddressRegistry.Api.Legacy.AddressMatch.Matching
-                    results.AddMunicipalitiesByPostalCode(municipalitiesGroup.Select(municipalityPostCode => municipalityPostCode.Item2), municipalitiesGroup.Key);
+            var municipalitiesByPostalCode = FilterByPartOfMunicipality(municipalities, results)
+                .GroupBy(municipalityPostCode => municipalityPostCode.Item1);
 
-                if (!results.Any())
-                {
-                    var municipalitiesByFuzzyMatch = municipalities.Where(g => g.DefaultName.Value.EqFuzzyMatch(results.Query.MunicipalityName, _config.FuzzyMatchThreshold));
-                    results.AddMunicipalities(municipalitiesByFuzzyMatch);
-                }
-            }
+            foreach (var municipalitiesGroup in municipalitiesByPostalCode)
+                results.AddMunicipalitiesByPostalCode(
+                    municipalitiesGroup
+                        .Select(municipalityPostCode => municipalityPostCode.Item2),
+                    municipalitiesGroup.Key);
 
-            //_telemetry.TrackGemeenteMatch(results.Query.Gemeentenaam, results.Query.Niscode, results.Query.Postcode, results.Count());
+            if (results.Any())
+                return results;
+
+            var municipalitiesByFuzzyMatch = municipalities
+                .Where(g => g
+                    .DefaultName.Value
+                    .EqFuzzyMatch(results.Query.MunicipalityName, _config.FuzzyMatchThreshold));
+
+            results.AddMunicipalities(municipalitiesByFuzzyMatch);
 
             return results;
         }
 
-        protected override bool IsValidMatch(AddressMatchBuilder matchResult) => matchResult.Any();
+        protected override bool? ShouldProceed(AddressMatchBuilder? matchResult)
+            => IsValidMatch(matchResult);
 
-        protected override bool ShouldProceed(AddressMatchBuilder matchResult) => IsValidMatch(matchResult);
+        protected override bool? IsValidMatch(AddressMatchBuilder? matchResult)
+            => matchResult.Any();
 
-        private IEnumerable<MunicipalityLatestItem> FilterByName(IEnumerable<MunicipalityLatestItem> municipalities, AddressMatchBuilder results)
+        protected override IReadOnlyList<TResult> BuildResultsInternal(AddressMatchBuilder? results)
+            => results
+                .Select(g => g.Municipality)
+                .Select(_mapper.Map)
+                .ToList();
+
+        private IEnumerable<MunicipalityLatestItem> FilterByName(
+            IEnumerable<MunicipalityLatestItem> municipalities,
+            AddressMatchBuilder? results)
         {
-            var municipalitiesByName = municipalities.Where(g => g.DefaultName.Value.EqIgnoreCase(results.Query.MunicipalityName)).ToList();
-            if (!string.IsNullOrEmpty(results.Query.MunicipalityName) && !municipalitiesByName.Any())
+            // TODO: Can't we skip this entire bit if there is no MunicipalityName query?
+
+            var municipalitiesByName = municipalities
```

Comment:
Can't we skip this entire bit if there is no `MunicipalityName` query?

Reply:
yep